### ISSUE TITLE
`keybase chat read` leaks unbounded amounts of memory

### ISSUE THREAD

**Issue by ToxicFrog** [keybase/client#18835]
Version: 4.1.0, 64-bit linux; I've also seen this on 3.x, but it's only now that I've been able to track down the cause.

Every time `keybase chat read` is called, the keybase daemon appears to leak a small amount of memory. In uses where it's called repeatedly, this means that the daemon's memory usage will grow without bound, until either it OOMs or the user notices and manually restarts it.

A simple and relatively quick reproduction is:

    while true; do
      keybase chat read <group> <channel> >/dev/null
    done

and when watch htop in another window.

On my machine this grows fairly slowly, perhaps 1MB/s, but this is still enough that within a few hours the system will become unusable.

Other `keybase chat` commands may exhibit the same behaviour, but `keybase chat read` is the only one I'm completely sure of at this point.

**Comment by maxtaco** [comment]
Thanks, we'll make a ticket.

**Comment by maxtaco** [comment]
How much memory on the system BTW?

**Comment by ToxicFrog** [comment]
16GB. When Keybase isn't running, it's usually about 50% in use.

**Comment by maxtaco** [comment]
We couldn't repro this, if you do a `keybase log send` we can take a look

**Comment by ToxicFrog** [comment]
Alright, reproducing it now to get some useful logs.

Is it still going to be useful if `keybase.service.log` doesn't exist? Because it shows up in `keybase status` but isn't on disk. The only log it's writing is `keybase.ek.log`. Nor is there anything in `journalctl` except for some "call cancelled" messages, matching up with when I was ^Cing `keybase chat read` while setting up the repro.

**Comment by maxtaco** [comment]
So that's an issue. I wonder if it's causing your memory problems since the logging is just getting endlessly buffered.

**Comment by ToxicFrog** [comment]
If every call to `keybase chat read` causes a log entry to be written, and logs are being stored in memory rather than written to disk, that would explain it neatly.

How do I start investigating this? strace keybase on startup to see what happens when it tries to `open` the log file?

**Comment by maxtaco** [comment]
Yes, that would work. You'd have to strace the persistent daemon process, and not the short-lived `keybase chat` process. And it could be the case that if you run it from your shell, you'll get different behavior than if you run it from systemd.  But it's worth a shot, sure.

**Comment by ToxicFrog** [comment]
I killed the daemon and started it from the shell, and...it doesn't even attempt to open `~/.cache/keybase/keybase.service.log`. It opens `keybase.ek.log` and `Keybase.app.log`, and of course there's lots of leveldb traffic and stuff in `.cache/keybase/attachments` and `.cache/keybase/avatars`. But it's not that it's failing to open the log, it doesn't look like it even tries.

It does show the steadily increasing memory usage.

If I run it with `--use-default-log-file` it properly opens and writes the log. Memory usage is still steadily creeping up even with the log file, though! I'll let it run for a while to see if it stabilizes and `log send` if not.

**Comment by ToxicFrog** [comment]
Log id `ffd28be2a85a82b84646e91c`

**Comment by ToxicFrog** [comment]
Also, here's the memory usage information from that repro:

```
(start memory monitoring and loop calling
 keybase chat read keybasefriends --channel=general --at-least=128)
DATE                              MEM% RSS
Thu Aug 22 10:21:27 EDT 2019 ----  2.0  319064
Thu Aug 22 11:21:29 EDT 2019 ----  9.0 1399112
Thu Aug 22 12:21:30 EDT 2019 ---- 19.5 3009604
Thu Aug 22 13:21:32 EDT 2019 ---- 28.2 4356752
(at this point I stopped the loop but kept monitoring memory usage)
Thu Aug 22 14:21:33 EDT 2019 ---- 23.5 3635684
Thu Aug 22 15:21:34 EDT 2019 ---- 22.6 3484976
Thu Aug 22 16:21:36 EDT 2019 ---- 22.0 3394292
Thu Aug 22 17:21:37 EDT 2019 ---- 22.3 3447740
Thu Aug 22 18:21:39 EDT 2019 ---- 22.0 3391772
Thu Aug 22 19:21:40 EDT 2019 ---- 23.1 3573004
Thu Aug 22 20:21:42 EDT 2019 ---- 22.3 3451544
Thu Aug 22 21:21:43 EDT 2019 ---- 26.1 4025876
```
So, once I stop hammering it with `chat read`, memory usage drops a bit, but not back to where it started, and as long as I'm calling read usage steadily increases.
